### PR TITLE
File-nally, MF4 Uploads are Here!

### PR DESCRIPTION
**Pull request by EvanHughes-dev**
Added mf4 file uploads and integrated with data_upload.py

### CODE REVIEW

File: TelemetrySite/client/src/DataUpload/DataUpload.jsx [hunk 1]
```diff
@@ -0,0 +1,166 @@
+/**
+ * Upload a mf4 file to the backend sever
+ * as well as the context id passed through
+ * the url. All file authentications are
+ * conducted on the backend
+ */
```
Comment (mjmagee991) [inline TelemetrySite/client/src/DataUpload/DataUpload.jsx]: Why is your line length so short?

File: TelemetrySite/client/src/DataUpload/DataUpload.jsx [hunk 2]
```diff
@@ -0,0 +1,166 @@
...
+
+    let lastProgress = 0; // Store previous progress percentage
+    let lastTimestamp = Date.now(); // Store previous timestamp
+
+    /**
+     * Helper function to convert seconds to hours and minutes
+     *
+     * @param {int} seconds - seconds to convert
+     * @return {string} formatted data
+     */
```
Comment (mjmagee991) [inline TelemetrySite/client/src/DataUpload/DataUpload.jsx:66]: Consider adding this type of documentation to all your functions here. Overall, I think this file is pretty readable, but the documentation is pretty sparse

File: TelemetrySite/client/src/DataUpload/DataUpload.jsx [hunk 3]
```diff
@@ -0,0 +1,166 @@
...
+      setProgressBar(
+        <div style={{ display: "flex", alignItems: "center" }}>
+          <progress value={currentProgress} style={{ marginRight: "10px" }} />
+          <div>{formatTime(estimatedTimeRemaining)}</div>
+        </div>
+      );
+
+      if (currentProgress >= 1 || currentProgress < 0) {
+        //stop calling to the backend
+        clearInterval(interval);
```
Comment (mjmagee991) [inline TelemetrySite/client/src/DataUpload/DataUpload.jsx:113]: How are you using `interval` inside the statement that is initializing it?

File: TelemetrySite/client/src/contextForm/ContextForm.jsx [hunk 4]
```diff
@@ -303,37 +350,58 @@ function ContextForm() {
     const MainElements = ContextJSONIdValues.MainElements;
 
     for (const key in MainElements) {
-      if (key !== "BikeConfig" || bikeIsCustom) {
+      if (key == "BikeConfig") {
```
Comment (mjmagee991) [inline TelemetrySite/client/src/contextForm/ContextForm.jsx]: Should this use `===`?

File: TelemetrySite/server/context_scripts/context_api.py [hunk 5]
```diff
@@ -4,7 +4,7 @@
 from psycopg2.sql import Identifier, SQL
 
 
-class Context(MethodView):
+class context_api(MethodView):
```
Comment (mjmagee991) [inline TelemetrySite/server/context_scripts/context_api.py]: Why are you using `snake_case` for a class name?

File: TelemetrySite/server/data_upload_scripts/data_upload.py [hunk 6]
```diff
@@ -110,8 +119,12 @@ def handle_data(files):
                 cur.executemany(sql, dataToExecuteMany)
             except TimeoutError:
                 print(len(dataToExecuteMany))
+
     conn.commit()
     conn.close()
+    # remove csv files from folder
+    for file in files:
+        os.remove(file)
```
Comment (mjmagee991) [inline TelemetrySite/server/data_upload_scripts/data_upload.py:127]: Are you 100% sure the user is going to want these files to be deleted?

File: TelemetrySite/server/data_upload_scripts/data_upload.py [hunk 7]
```diff
@@ -131,11 +144,13 @@ def format_data_bytes(bytes):
...
 
-## Main function to complete file conversion
-def main():
-    dotenv.load_dotenv("./credentials.env")
-    file_convert()
-
+# get the current progress of the file upload
+def get_progress(config_id):
+    if not config_id in progress_data:
+        return -1
```
Comment (mjmagee991) [inline TelemetrySite/server/data_upload_scripts/data_upload.py:150]: This function returns a failure or `None`. Did you mean to add more logic here?

File: TelemetrySite/server/data_upload_scripts/data_upload_api.py [hunk 8]
```diff
@@ -0,0 +1,92 @@
...
+                jsonify({"error": "Passed context id does not exits in database"}),
+                400,
+            )
+
+        # ensure the file actually contains a valid file name
+        if not file or file.name == "":
+            return jsonify({"error": "No file uploaded"}), 400
+
+        if self.file_type_check(file.filename):
+            # Secure name for best practice
```
Comment (mjmagee991) [inline TelemetrySite/server/data_upload_scripts/data_upload_api.py:65]: Change this comment to make it clearer what it actually means to "secure" a filename

File: TelemetrySite/server/server.py [hunk 9]
```diff
@@ -13,10 +14,23 @@
...
 # create views for url rules
-user_view = Context.as_view("context_api")
+user_view = context_api.as_view("context_api")
+
 app.add_url_rule(
     "/Context", view_func=user_view, methods=["GET", "PUT", "DELETE", "POST"]
 )
+user_view = data_upload_api.as_view("data_upload_api")
+# TODO Look at refactoring this later. I don't love
+# that data upload has the context takes in contextID
```
Comment (mjmagee991) [inline TelemetrySite/server/server.py]:
```suggestion
# that data upload takes in contextID
```

**Review by mjmagee991 — CHANGES_REQUESTED** [review]
A couple minor things

File: TelemetrySite/server/data_upload_scripts/data_upload.py (hunk 6 above)
Comment (EvanHughes-dev) [reply TelemetrySite/server/data_upload_scripts/data_upload.py:127]: This just delete the instance of the file saved by the program. The original file still exists on their computer or from where ever they uploaded it from, but the local copy saved by the server gets deleted to save space.

File: TelemetrySite/server/data_upload_scripts/data_upload_api.py (hunk 8 above)
Comment (EvanHughes-dev) [reply TelemetrySite/server/data_upload_scripts/data_upload_api.py:65]: Updated comment to describe what secure_filename does and what it changes in the name

File: TelemetrySite/client/src/contextForm/ContextForm.jsx (hunk 4 above)
Comment (EvanHughes-dev) [reply TelemetrySite/client/src/contextForm/ContextForm.jsx]: The code is correct as it is, but I just changed the logic to have it read better

File: TelemetrySite/server/context_scripts/context_api.py (hunk 5 above)
Comment (EvanHughes-dev) [reply TelemetrySite/server/context_scripts/context_api.py]: Switched all class names to `CamelCase`

File: TelemetrySite/client/src/DataUpload/DataUpload.jsx (hunk 2 above)
Comment (EvanHughes-dev) [reply TelemetrySite/client/src/DataUpload/DataUpload.jsx:66]: Updated documentation throughout file

File: TelemetrySite/server/server.py (hunk 9 above)
Comment (EvanHughes-dev) [reply TelemetrySite/server/server.py]: Updated comment

File: TelemetrySite/client/src/DataUpload/DataUpload.jsx (hunk 3 above)
Comment (EvanHughes-dev) [reply TelemetrySite/client/src/DataUpload/DataUpload.jsx:113]: This is the way intervals work. Despite being inside `SetInterval`, it is still able to access itself via `interval`, which prevents it from running.

File: TelemetrySite/server/data_upload_scripts/data_upload.py (hunk 7 above)
Comment (EvanHughes-dev) [reply TelemetrySite/server/data_upload_scripts/data_upload.py:150]: In this instance, no. If the key `config_id` isn't  in `progress_data`, then -1 is returned to stop the frontend from calling back. The frontend is setup to stop calling and end the progress bar if anything less than 0 is returned.

File: TelemetrySite/client/src/DataUpload/DataUpload.jsx (hunk 1 above)
Comment (EvanHughes-dev) [reply TelemetrySite/client/src/DataUpload/DataUpload.jsx]: updated lengths

**Review by mjmagee991 — APPROVED** [review]

**Pull request merged**